### PR TITLE
ADD Kernel Memory Allocator:

### PR DESCRIPTION
**Pull request by riblanc**
Adding a Kernel memory allocator based on a slab allocator.

Progress #82

### CODE REVIEW

**Comment by riblanc** [comment]
Sorry for da big pull request, whooooops :eyes:

File: src/memory/slab.zig [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+	pub fn free_object(self: *Self, obj: *usize) void {
+		const obj_addr = @intFromPtr(obj);
+
+		if (!self.is_obj_in_slab(obj))
+			@panic("Slab Allocator: Attempt to free a non kernel allocated object");
+
+		const index: u16 = @truncate((obj_addr - @intFromPtr(&self.data[0])) / self.header.cache.size_obj);
+		if (self.bitmap.get(index) catch .Free == .Free) @panic("Slab Allocator: Double free detected");
+		self.bitmap.set(index, Bit.Free) catch @panic("Slab Allocator: Bitmap corruption detected");
```
Comment (jgiron42) [inline src/memory/slab.zig:79]: remove panics

**Review by jgiron42 — CHANGES_REQUESTED** [review]

File: src/memory/fuzzer.zig [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+					self.add_chunk(ptr);
+					printk("\x1b[31malloc({d}) = 0x{x:0>8}\x1b[0m\n", .{size, @intFromPtr(ptr.ptr)});
+				} else if (self.size != 0) {
+					const chunk = self.rand.intRangeLessThan(usize, 0, self.size);
+					const ptr = self.allocs[chunk].ptr;
+					printk("\x1b[31mfree(0x{x:0>8})\x1b[0m\n", .{@intFromPtr(ptr)});
+					const sum = checksum(@intFromPtr(ptr));
+					for (self.allocs[chunk], 0..) |c, i| {
+						if (c != sum) {
+							printk("invalid checksum, expected {x}, got {x}{x}{x}{x} at {d}\n", .{sum, c, self.allocs[chunk][i+1], self.allocs[chunk][i+2], self.allocs[chunk][i+3], i});
```
Comment (riblanc) [inline src/memory/fuzzer.zig:43]:
```suggestion
							printk("invalid checksum, expected {x}, got {x} at {d}\n", .{sum, c, i});
```

File: src/memory.zig [hunk 3]
```diff
@@ -8,13 +8,18 @@ const multiboot = @import("multiboot.zig");
...
 const VirtualPageAllocator = @import("memory/virtual_page_allocator.zig").VirtualPageAllocator;
+const KernelMemoryAllocator = @import("memory/kernel_memory_allocator.zig").KernelMemoryAllocator;
 
 pub const VirtualPageAllocatorType = VirtualPageAllocator(PageFrameAllocator);
 
 pub var virtualPageAllocator : VirtualPageAllocatorType = .{};
 
 pub var pageFrameAllocator : PageFrameAllocator = .{};
 
+pub var global_cache: @import("memory/cache.zig").GlobalCache = undefined;
```
[the remark this answers is not in the document]
Comment (riblanc) [reply src/memory.zig]: Done 577bd61

File: src/memory/cache.zig [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+		order: u5
+	) Error!Cache {
+		var new = Cache{
+		.allocator = allocator,
+		.pages_per_slab = @as(usize, 1) << order,
+		// align the size of the object with usize
+		.size_obj = ft.mem.alignForward(usize, obj_size, @sizeOf(usize)),
+		};
+
+		// Compute the available space for the slab ((page_size * 2^order) - sise of slab header)
```
[the remark this answers is not in the document]
Comment (riblanc) [reply src/memory/cache.zig]: Done 9c7f569

File: src/memory/slab.zig (hunk 1 above)
Comment (riblanc) [reply src/memory/slab.zig]: Done 29f6c71

File: src/memory/kernel_memory_allocator.zig [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+		if (new_size < actual_size) return ptr[0..new_size];
+		var obj = try self.alloc(T, new_size);
+		@memcpy(obj[0..actual_size], ptr);
+		self.free(ptr);
+		return obj;
+	}
+
+	pub fn obj_size(_: *Self, ptr: anytype) !usize {
+		var pfd = global_cache.cache.get_page_frame_descriptor(@ptrCast(@alignCast(ptr)));
+		var slab: ?*Slab = if (pfd.next) |slab| @ptrCast(@alignCast(slab)) else null;
```
[the remark this answers is not in the document]
Comment (riblanc) [reply src/memory/kernel_memory_allocator.zig:85]: Done 29f6c71

File: src/memory/cache.zig [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+	pub fn init(allocator: *VirtualPageAllocatorType) !GlobalCache {
+		return .{ .cache = try Cache.init("global", allocator, @sizeOf(Cache), 0) };
+	}
+
+	pub fn create(self: *Self, name: []const u8, obj_size: usize, order: u5) !*Cache {
+		var cache: *Cache = @ptrCast(@alignCast(self.cache.alloc_one() catch |e| return e));
+
+		cache.* = Cache.init(
+			name, self.cache.allocator, obj_size, order
+		) catch @panic("Failed to initialize cache");
```
[the remark this answers is not in the document]
Comment (riblanc) [reply src/memory/cache.zig]: Done d92d258

**Review by jgiron42 — APPROVED** [review]

**Pull request merged**